File: pages/refactorings/replace-mutable-fields-with-lenses/01-rename-var/Spec.scala
//> using scala 3.3.4
//> using dep qa.hedgehog::hedgehog-core:0.14.0
//> using dep qa.hedgehog::hedgehog-runner:0.14.0
import hedgehog.*, hedgehog.core.*, hedgehog.runner.*

object Props extends Properties:
  def tests: List[Test] = List(
    property("upperVarName: Before == After", agrees),
    property("the hit is uppercased, misses pass through", hitAndMiss),
  )

  // A neutral tree, so one generated input feeds both Expr types.
  enum T:
    case TVar(name: String, ref: Int)
    case TApp(f: T, x: T)
    case TLam(bind: String, body: T)

  val genName: Gen[String] =
    Gen.alpha.list(Range.linear(0, 4)).map(_.mkString)
  def genT(depth: Int): Gen[T] =
    val leaf =
      for n <- genName; r <- Gen.int(Range.linear(-10, 10))
      yield T.TVar(n, r)
    if depth == 0 then leaf
    else
      Gen.choice1(
        leaf,
        for
          f <- genT(depth - 1)
          x <- genT(depth - 1)
        yield T.TApp(f, x),
        for b <- genName; body <- genT(depth - 1) yield T.TLam(b, body),
      )

  def toBefore(t: T): Before.Expr = t match
    case T.TVar(n, r)   => Before.Expr.EVar(Before.Var(n, r))
    case T.TApp(f, x)   => Before.Expr.EApp(toBefore(f), toBefore(x))
    case T.TLam(b, e)   => Before.Expr.ELam(b, toBefore(e))
  def fromBefore(e: Before.Expr): T = e match
    case Before.Expr.EVar(v)      => T.TVar(v.name, v.ref)
    case Before.Expr.EApp(f, x)  =>
      T.TApp(fromBefore(f), fromBefore(x))
    case Before.Expr.ELam(b, e)   => T.TLam(b, fromBefore(e))
  def toAfter(t: T): After.Expr = t match
    case T.TVar(n, r)   => After.Expr.EVar(After.Var(n, r))
    case T.TApp(f, x)   => After.Expr.EApp(toAfter(f), toAfter(x))
    case T.TLam(b, e)   => After.Expr.ELam(b, toAfter(e))
  def fromAfter(e: After.Expr): T = e match
    case After.Expr.EVar(v)     => T.TVar(v.name, v.ref)
    case After.Expr.EApp(f, x)  => T.TApp(fromAfter(f), fromAfter(x))
    case After.Expr.ELam(b, e)  => T.TLam(b, fromAfter(e))

  def agrees: Property =
    for t <- genT(4).forAll
    yield
      fromBefore(Before.upperVarName(toBefore(t)))
        ==== fromAfter(After.upperVarName(toAfter(t)))

  def hitAndMiss: Property =
    for
      n <- genName.forAll
      t <- genT(3).forAll
    yield
      val hit = After.upperVarName(After.Expr.EVar(After.Var(n, 0)))
        ==== After.Expr.EVar(After.Var(n.toUpperCase, 0))
      val miss =
        if tNoVars(t) then
          fromAfter(After.upperVarName(toAfter(t))) ==== t
        else Result.success
      hit and miss

  def tNoVars(t: T): Boolean = t match
    case T.TVar(_, _)   => false
    case T.TApp(f, x)   => tNoVars(f) && tNoVars(x)
    case T.TLam(_, b)   => tNoVars(b)
@main def spec(): Unit = SpecRunner.run(Props.tests)

File: pages/refactorings/replace-mutable-fields-with-lenses/01-rename-var/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Control.Monad (unless)
import System.Exit (exitFailure)
import Data.Char (toUpper)
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import qualified Before
import qualified After

-- A neutral tree, so one generated input feeds both Expr types.
data T = TVar String Int | TApp T T | TLam String T
  deriving (Eq, Show)

genName :: Gen String
genName = Gen.string (Range.linear 0 4) Gen.alpha

genT :: Int -> Gen T
genT 0 = TVar <$> genName <*> Gen.int (Range.linear (-10) 10)
genT depth = Gen.choice
  [ TVar <$> genName <*> Gen.int (Range.linear (-10) 10)
  , TApp <$> genT (depth - 1) <*> genT (depth - 1)
  , TLam <$> genName <*> genT (depth - 1)
  ]

toBefore :: T -> Before.Expr
toBefore (TVar n r) = Before.EVar (Before.Var n r)
toBefore (TApp f x) = Before.EApp (toBefore f) (toBefore x)
toBefore (TLam b e) = Before.ELam b (toBefore e)

fromBefore :: Before.Expr -> T
fromBefore (Before.EVar v)  = TVar (Before.vName v) (Before.vRef v)
fromBefore (Before.EApp f x) = TApp (fromBefore f) (fromBefore x)
fromBefore (Before.ELam b e) = TLam b (fromBefore e)

toAfter :: T -> After.Expr
toAfter (TVar n r) = After.EVar (After.Var n r)
toAfter (TApp f x) = After.EApp (toAfter f) (toAfter x)
toAfter (TLam b e) = After.ELam b (toAfter e)

fromAfter :: After.Expr -> T
fromAfter (After.EVar v)   = TVar (After.vName v) (After.vRef v)
fromAfter (After.EApp f x) = TApp (fromAfter f) (fromAfter x)
fromAfter (After.ELam b e) = TLam b (fromAfter e)

prop_agrees :: Property
prop_agrees = property $ do
  t <- forAll (genT 4)
  fromBefore (Before.upperVarName (toBefore t))
    === fromAfter (After.upperVarName (toAfter t))

noVars :: T -> Bool
noVars (TVar _ _) = False
noVars (TApp f x) = noVars f && noVars x
noVars (TLam _ e) = noVars e

prop_hit_and_miss :: Property
prop_hit_and_miss = property $ do
  n <- forAll genName
  t <- forAll (genT 3)
  After.upperVarName (After.EVar (After.Var n 0))
    === After.EVar (After.Var (map toUpper n) 0)
  if noVars t
    then fromAfter (After.upperVarName (toAfter t)) === t
    else success

main :: IO ()
main = do
  ok <- checkParallel $ Group "Props"
    [ ("upperVarName: Before == After", prop_agrees)
    , ("the hit is uppercased, misses pass through", prop_hit_and_miss)
    ]
  unless ok exitFailure
